### PR TITLE
Using CSV file to upload exact ingredients

### PR DESCRIPTION
**Pull request by slawosz**
https://regulatorydelivery.atlassian.net/browse/COSBETA-1757
Most of the functionality is focused around `app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb`

### CODE REVIEW

**Review by nasirkhanbeis — APPROVED** [review]
Tested this on the review app, looking all good as expected. Slawosz has kindly fixed all the edge cases. looking good. over to business user testing.

File: cosmetics-web/Gemfile [hunk 1]
```diff
@@ -20,7 +20,8 @@ gem "devise-security"
 gem "elasticsearch", "7.17.1"
 gem "elasticsearch-model"
 gem "elasticsearch-rails"
-gem "govuk-design-system-rails", "~> 0.10.1"
+# gem "govuk-design-system-rails", "~> 0.10.1"
+gem "govuk-design-system-rails", git: "https://github.com/govuk-ruby/govuk-design-system-rails", branch: "add-option-for-radio-wrapper-class", require: "govuk_design_system"
```
Comment (slawosz) [inline cosmetics-web/Gemfile:24]: I will take care of this as soon as new version of gem is released.

File: cosmetics-web/lib/active_model/type/strict_float.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/object/try"
+
+# Copied from
+# activemodel/lib/active_model/type/float.rb
+#
+# Just want to prevent casting of "20-20" to 20.0
+# Only change to original Float type is one line in cast_value line - see git history for details
+module ActiveModel
```
Comment (slawosz) [inline cosmetics-web/lib/active_model/type/strict_float.rb:10]: This is currently not used. I will either use it or delete.

File: cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    attribute :component
+
+    validate :file_size_validation
+    validate :file_is_csv_file_validation
+    validate :file_is_not_empty_validation
+    validate :correct_ingredients_validation
+
+    # Original valid is reseting error messages.
+    # We have an edge case here where we are adding error messages after
+    # validation - during actuall ingredient creation
```
Comment (scruti) [inline cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb:18]:
```suggestion
    # validation - during actual ingredient creation
```

File: cosmetics-web/app/models/ingredient.rb [hunk 4]
```diff
@@ -44,6 +44,7 @@ class Ingredient < ApplicationRecord
 
   validates :inci_name, presence: true, ingredient_name_format: { message: :invalid }
   validates :inci_name, uniqueness: { scope: :component_id }, if: :validate_inci_name_uniqueness?
+  validates :inci_name, length: { maximum: ResponsiblePersons::Notifications::IngredientConcentrationForm::NAME_LENGTH_LIMIT }
```
Comment (scruti) [inline cosmetics-web/app/models/ingredient.rb:47]: Please change :raising_hand: 

This would invalidate around 500 already existing ingredients in production. When we introduced the form validation, we wanted to ensure that previously existing ingredients were still valid. That is why the length validation wasn´t here.



```suggestion
  validates :inci_name, length: { maximum: ResponsiblePersons::Notifications::IngredientConcentrationForm::NAME_LENGTH_LIMIT }
```
Another option is to restrict the validation only to be checked on creation.

Also, if we are sharing the constant, I would invert the dependency, so the `NAME_LENGTH_LIMIT` is defined in the model, not in the form.

File: cosmetics-web/app/views/help/csv.html.erb [hunk 5]
```diff
@@ -0,0 +1,594 @@
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+
+    <h1 class="govuk-heading-l govuk-!-margin-bottom-2">
+      <!--abbr>CSV</abbr> file uploads for ingredients &ndash; help-->
```
Comment (scruti) [inline cosmetics-web/app/views/help/csv.html.erb:5]: :scissors: 
```suggestion
```

File: cosmetics-web/app/views/help/csv.html.erb [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+
+    <details class="govuk-details govuk-!-margin-left-0">
+
+      <summary class="govuk-details__summary">
+        <span class="govuk-details__summary-text">
+          Show the <abbr>NPIS</abbr> tables 
+        </span>
+      </summary>
+
+      <div class="govuk-details__text">
```
Comment (scruti) [inline cosmetics-web/app/views/help/csv.html.erb:58]: This whole section with the NPIS ingredients tables is duplicated between here and `app/views/responsible_persons/notifications/components/build/_ingredients_npis_needs_to_know.html.erb`
Could we extract it to a shared partial?

File: cosmetics-web/db/schema.rb [hunk 7]
```diff
@@ -15,6 +15,7 @@
   enable_extension "citext"
   enable_extension "pgcrypto"
   enable_extension "plpgsql"
+  enable_extension "uuid-ossp"
```
Comment (scruti) [inline cosmetics-web/db/schema.rb:18]: Please change :raising_hand: 
These schema changes seem unrelated to the PR (may be caused by a different PG version?)

File: cosmetics-web/spec/features/submit/notification_wizard/ingredients/using_csv_file_spec.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
+require "rails_helper"
+
+RSpec.describe "Adding ingredients to components", :with_stubbed_antivirus, type: :feature do
```
Comment (scruti) [inline cosmetics-web/spec/features/submit/notification_wizard/ingredients/using_csv_file_spec.rb:3]:
```suggestion
RSpec.describe "Adding ingredients to components using a CSV file", :with_stubbed_antivirus, type: :feature do
```

File: cosmetics-web/spec/features/submit/notification_wizard/ingredients/using_csv_file_spec.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    answer_is_item_available_in_shades_with "No"
+    answer_what_is_physical_form_of_item_with "Liquid"
+    answer_what_is_product_contained_in_with "A typical non-pressurised bottle, jar, sachet or other package"
+    answer_does_item_contain_cmrs_with "No"
+    answer_item_category_with "Hair and scalp products"
+    answer_item_subcategory_with "Hair and scalp care and cleansing products"
+    answer_item_sub_subcategory_with "Shampoo"
+  end
+
+  scenario "Adding exact concentration ingredients to a product" do
```
Comment (scruti) [inline cosmetics-web/spec/features/submit/notification_wizard/ingredients/using_csv_file_spec.rb:26]:
```suggestion
  scenario "Adding exact concentration ingredients to a product using a CSV file" do
```

File: cosmetics-web/spec/fixtures/files/Exact_ingredients_duplicate_row.csv [hunk 10]
```diff
@@ -0,0 +1,4 @@
+Sodium,35,498-19-8,false
+Aqua,65,,false
+ethonol,23,497-19-8,true
+ethonol,23,497-19-8,true
```
Comment (scruti) [inline cosmetics-web/spec/fixtures/files/Exact_ingredients_duplicate_row.csv:4]:
```suggestion
ethonol,23,497-19-8,true

```

File: cosmetics-web/spec/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form_spec.rb [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+    f = Tempfile.new
+    f.write(csv)
+    f.rewind
+    Rack::Test::UploadedFile.new(f, "text/csv")
+  end
+
+  before do
+    component
+  end
+
```
Comment (scruti) [inline cosmetics-web/spec/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form_spec.rb:29]: I think we are missing the `describe "validations"` and `describe "#save_ingredients` at top levels grouping the different assertions over `valid?` and `save_ingredients`

File: cosmetics-web/spec/models/notification_spec.rb [hunk 12]
```diff
@@ -45,7 +45,7 @@
     let(:image_upload) { create(:image_upload, :uploaded_and_virus_scanned) }
 
     context "when no information is missing" do
-      let(:notification) { build(:draft_notification, image_uploads: [image_upload], components: [component]) }
+      let(:notification) { create(:draft_notification, image_uploads: [image_upload], components: [component]) }
```
Comment (scruti) [inline cosmetics-web/spec/models/notification_spec.rb:48]: :thinking:  ?

File: cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    rescue IngredientCanNotBeSavedError => e
+      errors.add(:file, e.message)
+      false
+    end
+
+  private
+
+    def file_size_validation
+      if file_too_large?
+        errors.add(:file, "The selected file must be smaller than 15KB")
```
Comment (scruti) [inline cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb:56]: Since we have error messages defined in the locales `en.yml` file, should we move all the error messages there?

File: cosmetics-web/spec/models/notification_spec.rb (hunk 12 above)
Comment (slawosz) [reply cosmetics-web/spec/models/notification_spec.rb:48]: As we are using now `reload` in accept and submit validator, the record have to exists.

File: cosmetics-web/app/models/ingredient.rb (hunk 4 above)
Comment (slawosz) [reply cosmetics-web/app/models/ingredient.rb:47]: I will do `create` validation.

File: cosmetics-web/spec/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form_spec.rb (hunk 11 above)
Comment (slawosz) [reply cosmetics-web/spec/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form_spec.rb:29]: If you have idea maybe you can create a PR?

Comment (scruti) [reply cosmetics-web/spec/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form_spec.rb:29]: Is just adding the standard top level
```
describe "#method" do
...
end
``` 
for each of the subjects under test (`valid` and `save_ingredients` from what I see in the assertions.

File: cosmetics-web/spec/models/notification_spec.rb (hunk 12 above)
Comment (scruti) [reply cosmetics-web/spec/models/notification_spec.rb:48]: Thanks for clarifying!

File: cosmetics-web/.rubocop.yml [hunk 14]
```diff
@@ -119,3 +123,12 @@ RSpec/FilePath:
...
     - spec/support/**/*
+      
+RSpec/LeakyConstantDeclaration:
+  Exclude:
+    - spec/lib/active_model/type/strict_float_spec.rb
+      
+Lint/ConstantDefinitionInBlock:
+  Exclude:
+    - spec/lib/active_model/type/strict_float_spec.rb
+
```
Comment (scruti) [inline cosmetics-web/.rubocop.yml:134]:
```suggestion
```

**Review by scruti — APPROVED** [review]

File: cosmetics-web/app/controllers/responsible_persons/notifications/components/build_controller.rb [hunk 15]
```diff
@@ -313,6 +328,26 @@ def update_frame_formulation
     end
   end
 
+  def update_upload_ingredients_file
+    # if the params are empty and ingredients file already exist, we can proceed to next step
+    #   jump_to_step :want_to_add_another_ingredient, ingredients_uploaded: true
```
Comment (scruti) [inline cosmetics-web/app/controllers/responsible_persons/notifications/components/build_controller.rb:333]: Little bit confused by this comment, bellow it seems to be jumping to a different place:
` return jump_to_step :select_ph_option`

File: cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+    def valid?
+      component.ingredients.delete_all
+      return false if errors.present?
+
+      super
+    rescue ActiveRecord::StatementInvalid
+      errors.add(:file, "File has incorrect characters. Please check and try again")
+      false
+    end
+
```
Comment (scruti) [inline cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb:29]: Suggestions:
- This seems a private method if I did read the PR correctly. The only method called from outside is `save_ingredients` ?
- `component.ingredients.delete_all` called from `valid?` --> This seems better fitted for `save_ingredients`.

```suggestion
```

File: cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    def save_ingredients
+      return false unless valid?
+
+      # Despite validations above, there might be rare case when ingredient can not be saved, eg:
+      # * there will be duplicated ingredient in the file
+      # * between validation and creation ingredient will be created (very rare edge case)
+      # * difference between ingredient form and model validations
+      # * internal ActiveRecord issue
+      ActiveRecord::Base.transaction do
+        @ingredients.each_with_index do |ingredient, i|
```
Comment (scruti) [inline cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb:39]: Suggestion: Moving the deletion of the components existing ingredients from `save?` to `save_ingredients`.

Placed it inside the transaction to avoid a situation where the ingredient creation fails, but the component current ingredients are deleted anyways. 

```suggestion
      # Despite validations above, there might be rare case when ingredient can not be saved, eg:
      # * there will be duplicated ingredient in the file
      # * between validation and creation ingredient will be created (very rare edge case)
      # * difference between ingredient form and model validations
      # * internal ActiveRecord issue
      ActiveRecord::Base.transaction do
        component.ingredients.delete_all
        return false unless valid?
              
        @ingredients.each_with_index do |ingredient, i|
```

**Review by scruti — COMMENTED** [review]
Left some code suggestions in the `BulkIngredientUploadForm`.

File: cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+      ingredient.component = component
+      ingredient
+    end
+
+    def duplicated_ingredients_in_file?
+      names = []
+      ingredients_from_csv&.each_with_index do |row, i|
+        name = row[0]
+        if names.include? name
+          # errors.add(:file, "The file has error in row #{i + 1}")
```
Comment (scruti) [inline cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb:141]:
```suggestion
```

**Review by scruti — APPROVED** [review]

**Review by nasirkhanbeis — APPROVED** [review]
Re-tested part 2 changes, help text is now updated as per the ticket and new behaviour pre-select sub radio looks good and it wipes off the data when user choose to enter manually
